lib/cli/src/Cardano/CLI.hs

```diff
@@ -409,25 +419,94 @@ hoistKeyScheme eta s = CliKeyScheme
     , inspect = eta . inspect s
     }
 
+-- | Bidirectional encoding of data that allows specialization.
+--
+-- When decoding, the correponding, potentially specialized encoder is returned
+-- along with the result:
+-- >>> let (Codec _ decode) = ...
+-- >>> (x, encodeBack) <- decode
+--
+-- This allows us to have a @Codec@ for /both/ hex and bech32, and makes sure
+-- that we always encode the result back in the same encoding as we first got.
+--
+-- NOTE: Only supports one step of specialization. Could probably be fixed
+-- easily, but not needed now.
+data Codec m a b = Codec (a -> m (b, b -> m a)) (b -> m (a, a -> m b))
+
+-- | A @Category@ instance for @Codec@ allows us to use the @(.)@ operator.
+instance Monad m => Cat.Category (Codec m) where
+    id = Codec (\x -> return (x, return)) (\x -> return (x, return))
+    (.) = compose
+
+-- | A simple @Codec@ that does not specialize.
+simpleCodec :: Functor m => (a -> m b) -> (b -> m a) -> Codec m a b
+simpleCodec encode decode = Codec (fmap (,decode) . encode) (fmap (,encode) . decode)
+
+inverse :: Codec m a b -> Codec m b a
+inverse (Codec f g) = Codec g f
+
+compose :: Monad m => Codec m b c -> Codec m a b -> Codec m a c
+compose (Codec e1 d1) (Codec e2 d2) = Codec encode decode
+  where
+    encode x = do
+        (x', f) <- e2 x
+        (x'', g) <- e1 x'
+        return (x'', g >=> f)
+    decode x = do
+        (x', f) <- d1 x
+        (x'', g) <- d2 x'
+        return (x'', g >=> f)
+
 -- | Pair of functions representing the bidirectional transformation between
 -- a @XPrvOrXPub@ and its hex-encoded form.
-keyHexCodec :: (XPrvOrXPub -> Either String Text, Text -> Either String XPrvOrXPub)
-keyHexCodec = (encode, decode)
+keyHexCodec :: Codec (Either String) ByteString Text
+keyHexCodec = simpleCodec encode decode
+  where
+    encode :: ByteString -> Either String Text
+    encode = return . T.pack . B8.unpack . hex
+
+    decode :: Text -> Either String ByteString
+    decode txt = fromHex $ B8.pack $ T.unpack . T.strip $ txt
+      where
+        fromHex = left (const "Invalid hex.")
+            . convertFromBase Base16
+
+keyBech32Codec :: Codec (Either String) ByteString Text
+keyBech32Codec = simpleCodec encode decode
+  where
+    decode :: Text -> Either String ByteString
+    decode t = do
+        (hrp, dp) <- left show $ Bech32.decodeLenient t
+        bytes <- maybe (Left "todo") Right $ Bech32.dataPartToBytes dp
+        case hrp of
+            h | h == xpubHrp -> return bytes
+            h | h == xprvHrp -> return bytes
+            _ -> Left "unrecognized Bech32 Human Readable Part"
+
+    encode :: ByteString -> Either String Text
+    encode bs =
+        case BS.length bs of
+            96 -> return $ bech32 xprvHrp bs
```

Comment:
I think this can now be re-written as ` Codec (Either String) AXPrvOrXPub Text` which would remove the need to case-switch on the magic lengths here. It would also remove the need for codec composition — I think